forge/ee/lib/projectComms/index.js

```diff
@@ -62,41 +62,59 @@ module.exports.init = function (app) {
             'sub',
             { topic: /^ff\/v1\/([^/]+)\/p\/([^/]+)\/out\/[^/]+($|\/.*$)/, verify: 'checkDeviceIsAssigned' }
         )
-        // Receive messages sent to this project
+        // Receive messages sent to this device
         // - ff/v1/<team>/p/<project>/in/+/#
         app.comms.aclManager.addACL(
             'device',
             'sub',
-            { topic: /^ff\/v1\/([^/]+)\/p\/([^/]+)\/in\/[^/]+($|\/.*$)$/, verify: 'checkDeviceAssignedToProject' }
+            { topic: /^ff\/v1\/([^/]+)\/p\/([^/]+)\/in\/[^/]+($|\/.*$)$/, verify: 'checkDeviceCanAccessProject' }
         )
-        // Receive link-call response messages sent to this project
+        // Receive link-call response messages sent to this device (instance owned)
         // - ff/v1/<team>/p/<project>/res/+/#
         app.comms.aclManager.addACL(
             'device',
             'sub',
-            { topic: /^ff\/v1\/([^/]+)\/p\/([^/]+)\/res\/[^/]+($|\/.*$)$/, verify: 'checkDeviceAssignedToProject' }
+            { topic: /^ff\/v1\/([^/]+)\/p\/([^/]{11,})\/res\/[^/]+($|\/.*$)$/, verify: 'checkDeviceCanAccessProject' }
+        )
+        // Receive link-call response messages sent to this device (app owned)
+        // - ff/v1/<team>/p/<appid>/res/+/#
+        app.comms.aclManager.addACL(
+            'device',
+            'sub',
+            { topic: /^ff\/v1\/([^/]+)\/p\/([^/]{10})\/res\/[^/]+($|\/.*$)$/, verify: 'checkDeviceAssignedToApplication' }
         )
 
-        // Send message to other project
-        // - ff/v1/<team>/p/+/in/+/#
+        // Send message to specific project
+        // A device ID is exactly 10 chars long, so any topic with an ID longer than that
```

Comment:
It's a bit hard to figure out due to the way the git diff is showing the changes - but I'd expect any future support for a publish targeting a device to be on a `/d/` topic, not a `/p/` topic - so not sure this reliance on the length of this topic component is something we should rely on.

It would be helpful to a have a summary of the topics/rules being introduced in the PR description for ease of reference. I haven't got to the other related PRs to know if you've documented them elsewhere, but as this is the PR dealing with changing the platform ACLs, it would be handy to have here.

Reply:
It was an absolute melon twister Nick 😄 

I will attempt to answer the your questions in a meaningful manner (I am also working on suitable documentation for future reference)


>  but I'd expect any future support for a publish targeting a device to be on a /d/ topic, not a /p/ topic 

Agreed and considered and will likely be how it is handled for instance/device → device, but at this point in time, it is not supported and is in effect interlocked by insisting on the topic being 11 or more characters.  I suspect I will remove these once I have a working schema for instance/device → device



However, it was necessary to differentiate a PUB originating from an instance device vs PUB originating from an app assigned device in some scenarios and therefore an alternate verification routine was needed

Consider these

## "Send message to specific project"
### instance assigned device publishes direct to a project
1. The device has access to a `projectId` and can form the topic `ff/v1/7N152GxG2p/p/23d79df8-183c-4104-aa97-8915e1897326/in/a/b` 
    * this is the topic it uses to PUB direct to project `23d79df8-183c-4104-aa97-8915e1897326`
2. The receiving projects' Project-In node subscribes to `a/b` which equates to `ff/v1/7N152GxG2p/p/23d79df8-183c-4104-aa97-8915e1897326/in/a/b`

### application assigned device publishes direct to a project
1. As the Project In Node will subscribe to `ff/v1/7N152GxG2p/p/23d79df8-183c-4104-aa97-8915e1897326/in/a/b` the device must also publish on this topic

In summary - the `/p/` remains even though it is a device publishing


## "Broadcast a message"
### instance assigned device
1. The device has access to a `projectId` and can form the topic `ff/v1/7N152GxG2p/p/23d79df8-183c-4104-aa97-8915e1897326/out/a/b` 
    * this is the topic it uses to PUB broadcasts `23d79df8-183c-4104-aa97-8915e1897326`

### Project In subscription
3. A receiving projects' Project-In node subscribes to `a/b` which equates to `ff/v1/7N152GxG2p/p/+/out/a/b`
4. Therefore a device (both instance and app owned) Project-In node must also subscribe to `ff/v1/7N152GxG2p/p/+/out/a/b`

### application assigned device
1. The device does not have a `projectId`, only its own `deviceId` or the `applicationId`
2. Since the code already knows which device is which, I supplant the `projectId` with the  `applicationId`
5. This means we get the topic  `ff/v1/7N152GxG2p/p/1234567890/out/a/b`  (where `1234567890` is the app id)
6. This means we cannot use the same verification function since we are not verifying a project ID but an application ID

In summary - the `/p/` remains even though it is an application device publishing

To get over this hump, I chose to leave the Project In node as-is and differentiate the verification logic path by the `/ID/` provided.


The alternative solution I considered was to modify the Project-In node to either subscribe to both `ff/v1/7N152GxG2p/p/...` and `ff/v1/7N152GxG2p/d/...` OR subscribe to `ff/v1/7N152GxG2p/+/...`


---

Tomorrow, I hope to have a clearer picture of the direct instance/device → device and a full summary of current existing condition, this PR condition and our final stance.  

I think it may be prudent to hang fire and evaluate this before any merges?


---

PS: I appreciate that ↑ wasn't entirely as clear as I had hoped!  Imagine how my head was noodled doing the work ;)